src/main/java/com/streamr/client/dataunion/DataUnionClient.java

```diff
@@ -0,0 +1,320 @@
+                    return null;
+                return signatures;
+            }
+        };
+        return (BigInteger) waitForCondition(affirmationsChange, sleeptime, timeout);
+    }
+
+    public BigInteger waitForSidechainBalanceChange(BigInteger initialBalance, String balanceAddress, long sleeptime, long timeout) throws Exception {
+        return waitForErc20BalanceChange(initialBalance, sidechainToken().getContractAddress(), balanceAddress, sidechain, sleeptime, timeout);
+    }
+
+    public BigInteger waitForMainnetBalanceChange(BigInteger initialBalance, String balanceAddress, long sleeptime, long timeout) throws Exception {
+        return waitForErc20BalanceChange(initialBalance, mainnetToken().getContractAddress(), balanceAddress, mainnet, sleeptime, timeout);
+    }
+
+    public Boolean waitForMainnetTx(String txhash, long sleeptime, long timeout) throws Exception {
+        return waitForTx(mainnet, txhash, sleeptime, timeout);
+    }
+
+    public Boolean waitForSidechainTx(String txhash, long sleeptime, long timeout) throws Exception {
+        return waitForTx(sidechain, txhash, sleeptime, timeout);
+    }
+
+
+
+    /**
+     * port an AMB message that has required # signatures.
+     *
+     *
+     * @param msgHash AMB message hash
+     * @param collectedSignatures
+     * @param cred
+     * @throws Exception
+     *
+     * from foreign AMB hasEnoughValidSignatures(): Need to construct this binary blob:
+    * First byte X is a number of signatures in a blob,
+    * next X bytes are v components of signatures,
+    * next 32 * X bytes are r components of signatures,
+    * next 32 * X bytes are s components of signatures.
+    java byte is signed, so we use short to support 0-255
+
+     */
+
+    protected void portTxToMainnet(Bytes32 msgHash, short collectedSignatures, Credentials cred) throws Exception {
+        if(collectedSignatures > 255)
+            throw new UnsupportedOperationException("collectedSignatures cannot be greater than 255");
+
+        DynamicBytes message = sidechainAMB().message(msgHash).send();
+        byte[] signatures = new byte[1 + (65*collectedSignatures)];
+        signatures[0] = (byte) collectedSignatures;
+        //collect signatures one by one from sidechain, add to blob
+        for(short i = 0; i < collectedSignatures; i++){
+            Sign.SignatureData sig = fromBytes65(sidechainAMB().signature(msgHash, new Uint256(i)).send().getValue());
+            signatures[i+1] = sig.getV()[0];
+            System.arraycopy(sig.getR(), 0, signatures, 1+collectedSignatures+(i*32), 32);
+            System.arraycopy(sig.getS(), 0, signatures, 1+(collectedSignatures*33)+(i*32), 32);
+        }
+        mainnetAMB(cred).executeSignatures(message, new DynamicBytes(signatures)).send();
+    }
+
+    public void portTxsToMainnet(EthereumTransactionReceipt withdrawalTransaction, String mainnetSenderPrivateKey) throws Exception {
```

Comment:
Are these portTxsToMainnet methods targeted for application developers (the users of the streamr-java-client library)? If not, is it possible to change the methods to private/protected?

Reply:
yes for end users. must be public